editoast/src/core/simulation.rs

```diff
@@ -67,52 +69,135 @@ pub struct PhysicsRollingStock {
-        let towed_rolling_stock_mass = towed_rolling_stock.as_ref().map(|trs| trs.mass);
-        let mass = params
+        let towed_rolling_stock_mass = self
+            .towed_rolling_stock
+            .as_ref()
+            .map(|trs| trs.mass)
+            .unwrap_or(0.0);
+        let mass = self
             .total_mass
-            .unwrap_or_else(|| traction_engine_mass + towed_rolling_stock_mass.unwrap_or(0.0))
-            .round() as u64;
+            .unwrap_or(traction_engine_mass + towed_rolling_stock_mass);
+        mass.round() as u64
+    }
 
-        let max_speed = f64::min(
-            traction_engine.max_speed,
-            params.max_speed.unwrap_or(traction_engine.max_speed),
-        );
+    pub fn compute_rolling_resistance(&self, traction_engine: &RollingStock) -> RollingResistance {
+        if let (Some(towed_rolling_stock), Some(total_mass)) =
+            (self.towed_rolling_stock.as_ref(), self.total_mass)
+        {
+            let traction_engine_rr = &traction_engine.rolling_resistance;
+            let towed_rs_rr = &towed_rolling_stock.rolling_resistance;
+            let traction_engine_mass = traction_engine.mass; // kg
+
+            let mass_carriage = total_mass - traction_engine_mass; // kg
+
+            let rav_a_te = traction_engine_rr.A * 1000.0; // convert from kN to N
+            let rav_b_te = traction_engine_rr.B * 1000.0 * 3.6; // convert from kN/(km/h) to N/(m/s)
+            let rav_c_te = traction_engine_rr.C * 1000.0 * 3.6 * 3.6; // convert from kN/(km/h)² to N/(m/s)²
+
+            let rav_a_towed = towed_rs_rr.A * 1e-2 * mass_carriage; // convert from daN/t to N
+            let rav_b_towed = towed_rs_rr.B * 1e-2 * mass_carriage * 3.6; // convert from (daN/t)/(km/h) to N/(m/s)
+            let rav_c_towed = towed_rs_rr.C * 1e-2 * mass_carriage * 3.6 * 3.6; // convert from (daN/t)/(km/h)² to N/(m/s)²
+
+            let rav_a = rav_a_te + rav_a_towed;
+            let rav_b = rav_b_te + rav_b_towed;
+            let rav_c = rav_c_te + rav_c_towed;
+
+            RollingResistance::new(
+                traction_engine_rr.rolling_resistance_type.clone(),
+                rav_a,
+                rav_b,
+                rav_c,
+            )
+        } else {
+            traction_engine.rolling_resistance.clone()
+        }
+    }
+}
 
-        let comfort_acceleration =
-            compute_comfort_acceleration(&traction_engine, &towed_rolling_stock);
-        let startup_acceleration =
-            compute_startup_acceleration(&traction_engine, &towed_rolling_stock);
-        let inertia_coefficient =
-            compute_inertia_coefficient(&traction_engine, &towed_rolling_stock, params.total_mass);
-        let rolling_resistance =
-            compute_rolling_resistance(&traction_engine, &towed_rolling_stock, params.total_mass);
+impl PhysicsRollingStock {
+    pub fn new(traction_engine: RollingStock, physics_consist: PhysicsConsist) -> Self {
```

Comment:
Why separating consist from rolling stock. A consist contains a rolling stock and towed rolling stock. I really think `PhysicsRollingStock` should completely disappear in favor of `PhysicsConsist`.

Reply:
Thank you very much for this refactorization, it looks much better to me. Hope you agree 😄